src/claim_signer/GetClaimSigner.ts

```diff
@@ -0,0 +1,151 @@
+  readonly createClaim: (type: ClaimType, claimAttributes: object, context: ClaimContext) => Claim
+}
+
+export const getClaimSigner = () => {
+  const jsonld = JSONLD()
+
+  const getDataDocumentLoader = (jsonld: any) => {
+    const nodeDocumentLoader = jsonld.documentLoaders.node({ usePromise: true })
+
+    return async (url: string, callback: (error: any, data: any) => any) => {
+      const parsedData = ParseDataUrl(url)
+      if (parsedData) {
+        const publicKey = {
+          id: url,
+          type: 'Ed25519VerificationKey2018',
+          owner: url,
+          publicKeyBase58: parsedData.data,
+        }
+        return callback(null, {
+          contextUrl: ['https://w3id.org/security/v2'],
+          document: {
+            owner: {
+              id: url,
+              publicKey: [publicKey],
+            },
+            publicKey,
+          },
+        })
+      }
+      nodeDocumentLoader(url, callback)
+    }
+  }
+  jsonld.documentLoader = getDataDocumentLoader(jsonld)
+  const jsig = JSONLD_SIGS()
+  jsig.use('jsonld', jsonld)
+
+  const canonizeClaim = async (document: Claim, jsonld: any): Promise<string> => {
+    const contextualClaim = {
+      type: document.type,
+      '@context': {
+        ...DefaultClaimContext,
+        ...claimTypeDefaults[document.type],
+        ...document['@context'],
+      },
+      issuer: document.issuer,
+      issuanceDate: document.issuanceDate,
+      claim: document.claim,
+    }
+    return jsonld.canonize(contextualClaim)
+  }
+
+  const getClaimId = async (claim: Claim): Promise<string> => {
+    const canonizedClaim = await canonizeClaim(claim, jsonld)
+    const buffer = Buffer.from(canonizedClaim)
+    return crypto
+      .createHash('sha256')
+      .update(buffer)
+      .digest()
+      .toString('hex')
+  }
+
+  const isValidSignature = async (claim: Claim): Promise<boolean> => {
+    const results: any = await jsig.verify(claim, { checkNonce: cuid.isCuid })
+
+    return results.verified
+  }
+
+  const signClaimValidateClaimId = async (claim: Claim) => {
+    if (!claim.id) throw new IllegalArgumentException('Cannot sign a claim that has an empty .id field.')
+    if ((await getClaimId(claim)) !== claim.id)
+      throw new IllegalArgumentException('Cannot sign a claim whose id has been altered or generated incorrectly.')
+  }
+
+  const signClaimValidateSigningOptions = async (signingOptions: any) => {
+    if (signingOptions.creator === undefined || signingOptions.creator === null || signingOptions.creator === '')
+      throw new IllegalArgumentException('Cannot sign a claim with an invalid creator in the signing options.')
+  }
+
+  const signClaimValidateArgs = async (signingOptions: any, claim: Claim) => {
+    await signClaimValidateClaimId(claim)
+    await signClaimValidateSigningOptions(signingOptions)
+  }
+
+  const signingOptions = (privateKey: string) => ({
+    privateKeyBase58: privateKey,
+    algorithm: 'Ed25519Signature2018',
```

Comment:
the algorithm is hardcoded at the moment, Ed25519Signature2018 should be a default not hard coded. This is another area I think separating createClaim and signClaim can help with.

Reply:
+1 to using defaults rather than hard-coding wherever possible.

> This is another area I think separating createClaim and signClaim can help with.

In the old, old version of Po.et JS, `createClaim` was actually called `createSignedClaim`. I then moved on to thinking a claim can't be considered a claim if it isn't signed, just to simplify concepts.

Now that we're separating the responsibility of the key and signature to an Identity or issuer this may no longer hold true, though.

Just curious, what benefits would separating these functions bring? Do we need to do much with unsigned claims?

Reply:
We might not have any use for creating unsigned claims at the moment for our code base. But maybe others might, they might want to create claims, then sign them at a later time or a different part of their code base or even a different service.

The main benefit though is a simplified API. createClaim the way its set up now, has 4 arguments and it still does not provide a way for providing an algorithm override, so we need to adjust it further and I can't think of a great way to incorporate algorithm overrides into createClaim that don't make the API even more clunky.

I think separating createClaim and signClaim will greatly simplify the API, something like so

```
interface SigningOptions {
  readonly privateKey: string
  algorithm?: AgorithmTypes // This will default to Ed25519Signature2018 in signClaim
}
type signClaim = (signingOptions: SigningOptions) => (claim: Claim) => signedClaim: SignedClaim
type createClaim = (claimType: ClaimType) => (claimAttributes: ClaimAttributes) => claim: Claim
```

These are curried formats, they don't have to be, but I think these function will provide much more reusability when curried.

Reply:
Although if we do correctly separate the two above we could provide a function that combines those two like you originally named it by composing the other 2 functions. I personally prefer using the two functions above though.

```
interface CreateSignedClaimedConfig {
  readonly signClaimConfig: SignClaimConfig
  readonly createClaimConfig: CreateClaimConfig
}
type createSignedClaim = (config: CreateSignedClaimedConfig) => (claim: ClaimAttributes) => signedClaim: SignedClaim
```

Reply:
On another note, I think it might help things if we stop avoiding the name verifiable claim in our code.

In truth we are going from claim -> verifiableClaim -> signedVerifiableClaim

It would help our interfaces, API, documentation and communications a bunch IMO if we stopped referring to claims as claim attributes and stopped referring to verifiable claims as claims and called them what they are.

Reply:
I agree with not hardcoding the `Ed25519Signature2018` longterm, BUT the support is not there yet, which is why it is hardcoded. I need to do a bit more work before If open it up.

Reply:
And to repeat myself: the issuer IS part of the verifiableClaim, not just the signedVerifiableClaim. It is in the spec. For now, we are using `data:,publicKey` to get around the requirement that the signer (and issuer) MUST be valid URL values. Once we can get to url representations of our identity claims, then issuer can be a `po.et://` url that resolves to the public Key.

Reply:
Ah your right, just went and looked at verifiable claim again :) I forgot issuer was part of a verifiable claim, so my code above is incorrect, because createClaim is not returning a verifiable claim.

I might argue that ( just thinking this through for myself, please comment or we can discuss during our 1 on 1. )

**id does not need to be generated with the issuer**  
 - Signing is another wrapper of validation and isValidSignature uses getClaimId so signing prevents modification of the id, so issuer does not need to part of the id generation.
- This would mean a claim could be signed by multiple issuer ( not sure if this is good or bad? )


** id does require issuer during generation to prevent multiple issuers of the same claim id**  
- we entangle createClaim and signClaim or maybe we can instead move id generation to signClaim since we are tying those 2 things together anyway. createClaim just becomes a simple transform with no id generation.

Reply:
+1 to separating then, and also to using VerifiableClaim and Claim

Reply:
>  I need to do a bit more work before If open it up.

Then let's do that in a future PR if possible and keep it as it is now

Reply:
I was under the impression that signing and verification of verifiable claims is prescribed in the spec and encoded in the signature suites? Would changes like this make Po.et's verifiable claims incompatible with the broader verifiable claims ecosystem, and if so, is that a bad thing?

Reply:
Other important considerations:
* Would it delay mainnet launch
* How useful or necessary is it?
* Can we change our minds later without breaking changes?

Reply:
It is, however, the detail is our use of `data:,<publicKey>` for the url to resolve to the publicKey for verifying the signature. Until we are using REAL urls for creator/issuer, I haven't fully wired up the RSA portion.

Reply:
@kennylavender and I went through this so that it will not be a breaking change. We will always have a document resolver capable of managing `data:,` schemes for the creator. Using the ED22519 Signing algorithm is encased in all signed claims. Once we allow RSA support, those claims will be notated accordingly.